src/ir/item.rs

```diff
@@ -464,30 +473,53 @@ impl Item {
                         }
                     }
                 }
-                base
+                Some(base)
             }
             ItemKind::Var(ref var) => {
-                var.name().to_owned()
+                Some(var.name().to_owned())
             }
             ItemKind::Module(ref module) => {
                 module.name().map(ToOwned::to_owned)
-                    .unwrap_or_else(|| format!("_bindgen_mod{}", self.id()))
             }
         };
 
         let parent = ctx.resolve_item(self.parent_id());
         let parent_is_namespace = parent.is_module();
+
         if self.is_toplevel(ctx) || (parent_is_namespace && count_namespaces) {
+            let base_name = self.make_local_name(None, base_name, ctx);
             return ctx.rust_mangle(&base_name).into_owned();
         }
 
         // TODO: allow modification of the mangling functions, maybe even per
         // item type?
-        let parent = parent.canonical_name(ctx);
-        if parent.is_empty() {
-            base_name.to_owned()
-        } else {
-            format!("{}_{}", parent, base_name)
+        let parent_name = parent.canonical_name(ctx);
+        self.make_local_name(Some(parent_name), base_name, ctx)
+    }
+
+    fn make_local_name(&self,
+                       parent_name: Option<String>,
+                       base_name: Option<String>,
+                       ctx: &BindgenContext) -> String {
+        lazy_static! {
+            static ref RE_ENDS_WITH_BINDGEN_TY: Regex = Regex::new(r"_bindgen_ty_\d+$").unwrap();
+            static ref RE_ENDS_WITH_BINDGEN_MOD: Regex = Regex::new(r"_bindgen_mod_\d+$").unwrap();
+        }
+        let (re, kind) = match *self.kind() {
+            ItemKind::Module(..) => (&*RE_ENDS_WITH_BINDGEN_MOD, "mod"),
+            _ => (&*RE_ENDS_WITH_BINDGEN_TY, "ty"),
+        };
+        match (parent_name, base_name) {
```

Comment:
You need to take into account that `parent` might be empty, see `tests/headers/template_alias*`.